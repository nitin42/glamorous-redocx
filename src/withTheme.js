import React from 'react';
import PropTypes from 'prop-types';
import CHANNEL from './constants';

export default function withTheme(ComponentToTheme) {
  class ThemedComponent extends React.Component {
    state = { theme: {} };

    componentWillMount() {
      this.setState({ theme: this.context[CHANNEL].getState() });
    }

    componentDidMount() {
      if (this.context[CHANNEL]) {
        this.unsubscribe = this.context[CHANNEL].subscribe(this.setTheme);
      }
    }

    componentWillUnmount() {
      /* eslint-disable no-unused-expressions */
      this.unsubscribe;
    }

    setTheme = theme => this.setState({ theme });

    render() {
      /* eslint-disable react/jsx-filename-extension */
      return <ComponentToTheme {...this.props} {...this.state} />;
    }
  }

  ThemedComponent.contextTypes = {
    [CHANNEL]: PropTypes.object,
  };

  return ThemedComponent;
}
