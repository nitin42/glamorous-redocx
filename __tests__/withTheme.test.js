import React, {Component} from 'react'
import {render, mount} from 'enzyme'
import glamorous from '../src'
import withTheme from '../src/withTheme'
import ThemeProvider from '../src/themeProvider'
import CHANNEL from '../src/constants'

const getMockedContext = unsubscribe => ({
  [CHANNEL]: {
    getState: () => {},
    setState: () => {},
    subscribe: () => unsubscribe,
  },
})

const mockComp = glamorous.Text({})

test('renders a non-glamorous component with theme', () => {
  const CompWithTheme = withTheme(({theme: {fontSize}}) => (
    <glamorous.TEXT style={{fontSize}} />
  ))
  expect(
    render(
      <ThemeProvider theme={{fontSize: 10}}>
        <CompWithTheme />
      </ThemeProvider>,
    ),
  ).toMatchSnapshot()
})

test('theme properties updates get propagated down the tree', () => {
  class Parent extends Component {
    state = {
      fontSize: 10,
    }

    render() {
      return (
        <ThemeProvider theme={{fontSize: this.state.fontSize}}>
          <Child />
        </ThemeProvider>
      )
    }
  }

  const Child = withTheme(({theme: {fontSize}}) => <glamorous.TEXT style={{fontSize}} />)
  const wrapper = mount(<Parent />)
  expect(wrapper).toMatchSnapshot(`with theme prop of fontSize 10`)
  wrapper.setState({fontSize: 20})
  expect(wrapper).toMatchSnapshot(`with theme prop of fontSize 20`)
})

test('works properly with classes', () => {
  /* eslint-disable react/prefer-stateless-function */
  class Child extends Component {
    render() {
      const {theme: {fontSize}} = this.props
      return <glamorous.TEXT style={{fontSize}} />
    }
  }

  const ChildWithTheme = withTheme(Child)

  class Parent extends Component {
    state = {
      fontSize: 10,
    }

    render() {
      return (
        <ThemeProvider theme={{fontSize: this.state.fontSize}}>
          <ChildWithTheme />
        </ThemeProvider>
      )
    }
  }

  const wrapper = mount(<Parent />)
  expect(wrapper).toMatchSnapshot(`with theme prop of fontSize 10`)
})

test('unsubscribes from theme updates on unmount', () => {
  const Child = withTheme(() => <glamorous.TEXT />)
  const unsubscribe = jest.fn()
  const context = getMockedContext(unsubscribe)
  const wrapper = mount(<ThemeProvider theme={{}}><Child /></ThemeProvider>, {
    context,
  })
  wrapper.unmount()
  expect(unsubscribe).toHaveBeenCalled()
})
