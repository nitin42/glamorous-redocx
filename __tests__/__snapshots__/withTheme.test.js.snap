// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a non-glamorous component with theme 1`] = `
<text
  style="font-size:10px;"
/>
`;

exports[`with theme prop of fontSize 10 1`] = `
<Parent>
  <ThemeProvider
    theme={
      Object {
        "fontSize": 10,
      }
    }
  >
    <ThemedComponent>
      <Component
        theme={
          Object {
            "fontSize": 10,
          }
        }
      >
        <glamorous.Text
          style={
            Object {
              "fontSize": 10,
            }
          }
        >
          <TEXT
            style={
              Object {
                "fontSize": 10,
              }
            }
          />
        </glamorous.Text>
      </Component>
    </ThemedComponent>
  </ThemeProvider>
</Parent>
`;

exports[`with theme prop of fontSize 10 2`] = `
<Parent>
  <ThemeProvider
    theme={
      Object {
        "fontSize": 10,
      }
    }
  >
    <ThemedComponent>
      <Child
        theme={
          Object {
            "fontSize": 10,
          }
        }
      >
        <glamorous.Text
          style={
            Object {
              "fontSize": 10,
            }
          }
        >
          <TEXT
            style={
              Object {
                "fontSize": 10,
              }
            }
          />
        </glamorous.Text>
      </Child>
    </ThemedComponent>
  </ThemeProvider>
</Parent>
`;

exports[`with theme prop of fontSize 20 1`] = `
<Parent>
  <ThemeProvider
    theme={
      Object {
        "fontSize": 20,
      }
    }
  >
    <ThemedComponent>
      <Component
        theme={
          Object {
            "fontSize": 20,
          }
        }
      >
        <glamorous.Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          <TEXT
            style={
              Object {
                "fontSize": 20,
              }
            }
          />
        </glamorous.Text>
      </Component>
    </ThemedComponent>
  </ThemeProvider>
</Parent>
`;
